patches/server/0725-Improve-entity-damage-with-descriptive-return-type.patch

```diff
@@ -0,0 +1,1478 @@
+ 
+     @Override
+     public Brain<Zoglin> getBrain() {
+-        return super.getBrain();
++        return (Brain<Zoglin>) super.getBrain(); // Paper - decompile error - unchecked generic cast
+     }
+ 
+     protected void updateActivity() {
+diff --git a/src/main/java/net/minecraft/world/entity/monster/Zombie.java b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
+index bb3b932c57fd1e5b1517940c7602c7f4aeeaf17e..d95924a5e7eafbaf3efdf8d3bc79ee1bb16900f3 100644
+--- a/src/main/java/net/minecraft/world/entity/monster/Zombie.java
++++ b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
+@@ -310,12 +310,15 @@ public class Zombie extends Monster {
+     }
+     // Paper end
+ 
++    // Paper start - improved entity damage return value
+     @Override
+-    public boolean hurt(DamageSource source, float amount) {
+-        if (!super.hurt(source, amount)) {
+-            return false;
++    public EntityHurtResult tryToHurt(DamageSource source, float amount) {
++        final EntityHurtResult result = super.tryToHurt(source, amount);
++        if (!result.isHurt()) {
++            return result;
+         } else if (!(this.level instanceof ServerLevel)) {
+-            return false;
++            return EntityHurtResult.NOT_HURT;
++            // Paper end
+         } else {
+             ServerLevel worldserver = (ServerLevel) this.level;
+             LivingEntity entityliving = this.getTarget();
+@@ -352,7 +355,7 @@ public class Zombie extends Monster {
+                 }
+             }
+ 
+-            return true;
++            return EntityHurtResult.HURT; // Paper - improved entity damage return value
+         }
+     }
+ 
+diff --git a/src/main/java/net/minecraft/world/entity/monster/hoglin/Hoglin.java b/src/main/java/net/minecraft/world/entity/monster/hoglin/Hoglin.java
+index c510da19883d1aa79b2fc25e2d9c8f5cd8dd7bfa..e968787ebd66990aafd8f212bd15f8b0a31db727 100644
+--- a/src/main/java/net/minecraft/world/entity/monster/hoglin/Hoglin.java
++++ b/src/main/java/net/minecraft/world/entity/monster/hoglin/Hoglin.java
+@@ -97,19 +97,21 @@ public class Hoglin extends Animal implements Enemy, HoglinBase {
+ 
+     }
+ 
++    // Paper start - improved entity damage return value
+     @Override
+-    public boolean hurt(DamageSource source, float amount) {
+-        boolean bl = super.hurt(source, amount);
++    public EntityHurtResult tryToHurt(DamageSource source, float amount) {
++        final EntityHurtResult bl = super.tryToHurt(source, amount);
+         if (this.level.isClientSide) {
+-            return false;
++            return EntityHurtResult.NOT_HURT;
+         } else {
+-            if (bl && source.getEntity() instanceof LivingEntity) {
++            if (bl.isHurt() && source.getEntity() instanceof LivingEntity) {
+                 HoglinAi.wasHurtBy(this, (LivingEntity)source.getEntity());
+             }
+ 
+             return bl;
+         }
+     }
++    // Paper end
```

Comment:
Another out-of-place comment.

For the block this is closing, I would use single inline comments instead of a big block. They are more of one line changes, instead of changing a big block, or inserting a big block.